--- a/fs/kernfs/dir.c
+++ b/fs/kernfs/dir.c
@@ -1522,6 +1522,9 @@
 	struct kernfs_root *root = kernfs_root(kn);
 	struct kernfs_super_info *info;
 
+       if (!(kn->flags & KERNFS_INODE_INITED))
+               return;
+
 	lockdep_assert_held_read(&root->kernfs_supers_rwsem);
 
 	list_for_each_entry(info, &root->supers, node) {
--- a/fs/kernfs/inode.c
+++ b/fs/kernfs/inode.c
@@ -205,6 +205,10 @@
 
 	set_default_inode_attr(inode, kn->mode);
 	kernfs_refresh_inode(kn, inode);
+
+       down_write(&kernfs_root(kn)->kernfs_iattr_rwsem);
+       kn->flags |= KERNFS_INODE_INITED;
+       up_write(&kernfs_root(kn)->kernfs_iattr_rwsem);
 
 	/* initialize inode according to type */
 	switch (kernfs_type(kn)) {
--- a/include/linux/kernfs.h
+++ b/include/linux/kernfs.h
@@ -118,6 +118,7 @@
 	KERNFS_EMPTY_DIR	= 0x1000,
 	KERNFS_HAS_RELEASE	= 0x2000,
 	KERNFS_REMOVING		= 0x4000,
+       KERNFS_INODE_INITED     = 0x8000,
 };
 
 /* @flags for kernfs_create_root() */
